ifrs17/Import/Importers.ipynb

```diff
@@ -593,6 +571,100 @@
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public async static Task ValidateDataNodeStatesAsync(this IWorkspace workspace, Dictionary<string, DataNodeData> persistentDataNodeByDataNode)",
+                "\n{",
+                "\n    foreach(var importedDataNodeState in await workspace.Query<DataNodeState>().ToArrayAsync())",
+                "\n    {",
+                "\n        if(persistentDataNodeByDataNode.TryGetValue(importedDataNodeState.DataNode, out var currentPersistentDataNode))",
+                "\n        {",
+                "\n            if(importedDataNodeState.State < currentPersistentDataNode.State)",
+                "\n                ApplicationMessage.Log(Error.ChangeDataNodeState, importedDataNodeState.DataNode, ",
+                "\n                                                                  currentPersistentDataNode.State.ToString(), ",
+                "\n                                                                  importedDataNodeState.State.ToString());",
+                "\n",
+                "\n            if(importedDataNodeState.State == currentPersistentDataNode.State)",
+                "\n                await workspace.DeleteAsync<DataNodeState>(importedDataNodeState);",
+                "\n        }",
+                "\n    }",
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "markdown",
+            "source": [
+                "## Validate for Mandatory Aoc Steps"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public async static Task ValidateForMandatoryAocSteps(this IWorkspace workspace, IDataSet dataSet, HashSet<AocStep> mandatoryAocSteps)",
+                "\n{",
+                "\n    var excludedProperties = new[]{nameof(AocType), nameof(Novelty)};",
+                "\n    var missingAocStepsByIdentityProperties = (await workspace.Query<RawVariable>().ToListAsync())",
+                "\n                     .GroupBy(x => string.Join(\", \", x.ToIdentityString().Split(',').Where(y=> y != \" \" && !excludedProperties.Contains(y.Split(':')[0]))),",
+                "\n                              x => new AocStep(x.AocType, x.Novelty),",
+                "\n                              (properties, parsedAocSteps) => (properties, mandatoryAocSteps.Except(parsedAocSteps))",
+                "\n                             );",
+                "\n    foreach((var properties, var missingSteps) in missingAocStepsByIdentityProperties) ",
+                "\n        foreach(var missingStep in missingSteps)  ApplicationMessage.Log(Warning.MandatoryAocStepMissing, missingStep.AocType, missingStep.Novelty, properties);",
```

Comment:
I would log an Error. Why a Warning?

Reply:
This was chosen so by Daniel, with a reason that I don't know.

Reply:
ok, let's leave it for the time being.